lib/class-wp-rest-block-types-controller.php

```diff
@@ -0,0 +1,376 @@
+		$this->rest_base = 'block-types';
+		$this->registry  = WP_Block_Type_Registry::get_instance();
+	}
+
+	/**
+	 * Registers the routes for the objects of the controller.
+	 *
+	 * @see   register_rest_route()
+	 */
+	public function register_routes() {
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_items' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+					'args'                => $this->get_collection_params(),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/(?P<namespace>[a-zA-Z0-9_-]+)/(?P<name>[a-zA-Z0-9_-]+)',
+			array(
+				'args'   => array(
+					'name'      => array(
+						'description' => __( 'Block name', 'gutenberg' ),
+						'type'        => 'string',
+					),
+					'namespace' => array(
+						'description' => __( 'Block namespace', 'gutenberg' ),
+						'type'        => 'string',
+					),
+				),
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_item' ),
+					'permission_callback' => array( $this, 'get_item_permissions_check' ),
+					'args'                => array(
+						'context' => $this->get_context_param( array( 'default' => 'view' ) ),
+					),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Checks whether a given request has permission to read post block types.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|bool True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		if ( ! current_user_can( 'edit_posts' ) ) {
+			return new WP_Error( 'rest_cannot_view', __( 'Sorry, you are not allowed to manage block types.', 'gutenberg' ), array( 'status' => rest_authorization_required_code() ) );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Retrieves all post block types, depending on user context.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+		$data        = array();
+		$block_types = $this->registry->get_all_registered();
+
+		// Retrieve the list of registered collection query parameters.
+		$registered = $this->get_collection_params();
+		$namespace  = '';
+		if ( isset( $registered['namespace'] ) && ! empty( $request['namespace'] ) ) {
+			$namespace = $request['namespace'];
+		}
+
+		foreach ( $block_types as $slug => $obj ) {
+			$block_type = $this->prepare_item_for_response( $obj, $request );
+			if ( $namespace ) {
```

Comment:
Can we do the namespace check before preparing the item for response?

Reply:
This filters the request by namespace. So `?namespace=core`

Reply:
Right, I'm saying can we check if the block should be included before calling `prepare_item_for_response` so we don't prepare a block type if we are just going to discard it.